### PR TITLE
WIP - implement ksync support using subcommands for sync

### PR DESCRIPTION
**Pull request by ashahba**
fixes #{issue_num_here_if_applicable}

If you aren't done, please add the `wip` label until you're ready for review.

If you are adding a new template or mlt command/flag ensure that is is documented in the [docs/features.md](https://github.com/IntelAI/mlt/tree/master/docs/features.md) file.

## Reviewer Checklist

 - [ ] Do you understand it?
 - [ ] Is it correct?
 - [ ] Is it safe?
 - [ ] Is it legally compliant?
 - [ ] Is it robust?
 - [ ] Is it simple?
 - [ ] Is it tested?
 - [ ] Is it documented?
 - [ ] Is the style consistent?

See [the reviewer guide](docs/reviews.md) for more detail on each check.

### CODE REVIEW

**Comment by ashahba** [comment]
I still need to fix the unit tests.

**Comment by ashahba** [comment]
@s1113950 and @dmsuehir this has everyting that initially was provided by #223 plus more.
We just use subcommands for `sync`, like `mlt sync create`, `mlt sync reload` and `mlt sync delete`.

All new feedbacks should be addressed in a separate PR since this is way overdue now 🙂

File: mlt/commands/sync.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def action(self):
             self._create()
         elif self.args.get('reload'):
             self.__reload()
-        else:
+        elif self.args.get('delete'):
```
Comment (s1113950) [inline mlt/commands/sync.py]: This is the only other option in this series of commands; the `elif` isn't necessary

File: mlt/commands/sync.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def action(self):
             self._create()
         elif self.args.get('reload'):
             self.__reload()
```
Comment (s1113950) [inline mlt/commands/sync.py]: This function should be renamed to `self._reload()`? Did you mean to have 2 underscores here?

File: mlt/main.py [hunk 3]
```diff
@@ -53,19 +53,21 @@
...
-                            running pods which in turn restarts the containers
-                            every time changes are made to local template code.
-                            This command only needs to run once.
-                            'mlt reload' - to wake up the 'sync' agent after a
-                            local system reboot or long inactivity (default is
-                            1 hour) or any other activity that causes 'sync'
-                            agent to die.
-                            'mlt unsync' - to teardown syncing setup and stop
-                            syncing local code changes with remote pods.
+                            'sync' and it's subcommands 'reload' and 'delete'
```
Comment (s1113950) [inline mlt/main.py]: `its`*

File: tests/unit/commands/test_sync.py [hunk 4]
```diff
@@ -102,6 +104,40 @@ def test_sync_no_enable_sync(init_mock, open_mock, isfile_mock):
...
+    with catch_stdout() as caught_output:
+        with pytest.raises(SystemExit):
+            sync.action()
+        output = caught_output.getvalue()
+        expected_output = "This app is not initialized with " \
+                          "'--enable-sync' option"
+        assert expected_output in output
+
+
+def test_delete_sync_no_enable_sync(init_mock, open_mock, isfile_mock):
```
Comment (s1113950) [inline tests/unit/commands/test_sync.py]: This test and the one above it could be parametrized; only difference is the dict passed tot `SyncCommand`

File: tests/unit/commands/test_sync.py [hunk 5]
```diff
@@ -117,12 +153,42 @@ def test_sync_synced(init_mock, open_mock, isfile_mock, json_mock,
...
+
+    with catch_stdout() as caught_output:
+        with pytest.raises(SystemExit):
+            sync.action()
+        output = caught_output.getvalue()
+        expected_output = "This app has not been deployed yet"
+        assert expected_output in output
+
+
+def test_reload_sync_no_deploy(init_mock, open_mock, isfile_mock):
```
Comment (s1113950) [inline tests/unit/commands/test_sync.py]: This test and the one above it and below it can also be parametrized

File: tests/unit/commands/test_sync.py [hunk 6]
```diff
@@ -133,47 +199,196 @@ def test_sync_no_deploy(init_mock, open_mock, isfile_mock):
...
-    error_msg = "No rule to make target `sync-create'"
+    error_msg = "This app does not support the `mlt sync create` command"
     subprocess_mock.side_effect = CalledProcessError(
         returncode=2, cmd="make sync-create", output=error_msg)
-    sync_output = sync()
+    sync_output = sync(create=True)
+    assert error_msg in sync_output
+
+
+def test_sync_reload_not_in_makefile(init_mock, open_mock, isfile_mock,
```
Comment (s1113950) [inline tests/unit/commands/test_sync.py]: This test and the one below it can also be parametrized

File: mlt/commands/init.py [hunk 7]
```diff
@@ -79,6 +98,26 @@ def action(self):
...
+        k8s_template_specs = []
+        for filename in os.listdir(k8s_template_dir):
+            if fnmatch.fnmatch(filename, '*job.yaml'):
+                k8s_template_specs.append(os.path.join(
+                    k8s_template_dir, filename))
+
+        for filename in k8s_template_specs:
+            with open(filename, 'r+') as f:
+                filedata = f.read()
+                filedata = filedata.replace('#  ', '   ')
```
Comment (dmsuehir) [inline mlt/commands/init.py]: hmm... this means we can't have single `#` comments anywhere in the k8s-template yamls?  😬 Would it be possible to instead look for certain start/end comments and only uncomment between those known lines?  Otherwise, I can easily see someone else adding `# comment` somewhere in the yaml and us not remembering about this caveat.

File: mlt/commands/status.py [hunk 8]
```diff
@@ -51,6 +51,14 @@ def action(self):
                                              env=user_env,
                                              stderr=subprocess.STDOUT)
             print(output.decode("utf-8").strip())
+            if sync_helpers.get_sync_spec() is not None:
+                # format the string before passing it print so that the string
+                # is evaluated correctly
+                print("\nSYNC STATUS\n{}".format(
+                    'This app is being watched by Sync'))
+            else:
+                print("\nSYNC STATUS\n{}".format(
```
Comment (dmsuehir) [inline mlt/commands/status.py]: Is there a way for us to know if syncing was even enabled on this app to begin with?  It seems like if syncing was never enabled, we shouldn't show this, since syncing is not even an option.

File: mlt/commands/sync.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                # TODO: when we have a template updating capability, add a
+                # note recommending that he user update's their template to
+                # get the sync create command
+                print("This app does not support the `mlt sync create` "
+                      "command. No `sync-create` target was found in the "
+                      "Makefile.")
+            else:
+                print("Error while creating sync spec: {}".format(e.output))
+
+    def __reload(self):
```
Comment (dmsuehir) [inline mlt/commands/sync.py]: Why does `__reload()` have double underscores?

File: mlt/main.py [hunk 10]
```diff
@@ -50,6 +52,26 @@
                             use a namespace identical to username.
   --skip-crd-check          To avoid crd check during mlt init
                             [default: False].
+  --enable-sync             Without initializing templates with this option
+                            'sync' and it's subcommands 'reload' and 'delete'
```
Comment (dmsuehir) [inline mlt/main.py]: missing mention of 'create'?

File: tests/unit/commands/test_sync.py [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+    init_mock.return_value = mlt_config
+    json_mock.load.return_value = {"app_run_id": "123-456-789"}
+    get_sync_spec_mock.return_value = "hello-world"
+    expected_output = "Sync agent is restarted"
+    subprocess_mock.return_value.decode.return_value = expected_output
+    sync_output = sync(reload=True)
+    assert expected_output in sync_output
+
+
+def test_delete_not_synced(init_mock, open_mock, isfile_mock, json_mock,
```
Comment (dmsuehir) [inline tests/unit/commands/test_sync.py]: Also can parameterize this and test_reload_not_synced

File: README.md [hunk 12]
```diff
@@ -41,6 +42,10 @@ Prerequisites:
 - [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) (for the tensorboard template)
 - [openssh-server] Please install version specific to your OS
 
+Optional prerequisits:
```
[the remark this answers is not in the document]
Comment (ashahba) [reply README.md]: Yes I agree @dmsuehir 
If we want to make the list consistent in this PR, would you mind give me the full list and we just add them to this PR?

[the remark this answers is not in the document]
Comment (dmsuehir) [reply README.md]: I think it's this...

Things that are required in general:
- [Docker](https://docs.docker.com/install/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [python](https://www.python.org/downloads/)
- [pip](https://pip.pypa.io/en/stable/installing/)

Things that are required for certain functions or certain templates:
- [kubetail](https://github.com/johanhaleby/kubetail) (for `mlt logs` command)
- [Ksync](https://github.com/vapor-ware/ksync) (for `mlt sync` commands)
- [ks](https://github.com/ksonnet/ksonnet#install) (for the horovod template)
- [openssh-server] Please install version specific to your OS (for the horovod template)
- [TFJob operator](https://github.com/kubeflow/tf-operator#installing-the-tfjob-crd-and-operator-on-your-k8s-cluster) (for the distributed tensorflow templates)
- [PyTorch operator](https://github.com/kubeflow/pytorch-operator) (for the pytorch-distributed template)
- [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) (for the tensorboard template)

[the remark this answers is not in the document]
Comment (dmsuehir) [reply README.md]: oh, just noticed `openssh-server` is missing a link 😬

File: mlt/commands/deploy.py [hunk 13]
```diff
@@ -42,6 +43,9 @@ def __init__(self, args):
 
     def action(self):
         schema.validate()
+        if sync_helpers.get_sync_spec() is not None:
```
[the remark this answers is not in the document]
Comment (ashahba) [reply mlt/commands/deploy.py]: Oh, I guess the main reason was that if they do new deployment the old one is still being synced but I guess we should solve that in general and I'll remove this gate from this PR.

File: mlt/commands/init.py (hunk 7 above)
Comment (ashahba) [reply mlt/commands/init.py]: Not all comments.
Only the ones in this format:
`#` followed by `2 spaces` are replaced by `3 spaces`.
I'm open to suggestions.

File: mlt/commands/status.py (hunk 8 above)
Comment (ashahba) [reply mlt/commands/status.py]: Yes, let me see if I can fix that.

File: mlt/commands/init.py (hunk 7 above)
Comment (ashahba) [reply mlt/commands/init.py]: Yes I can do that.

Comment (dmsuehir) [reply mlt/commands/init.py]: Yeah, I was thinking of something like what Steven has.  What do you guys think?

I was just concerned that someone might inadvertently add a comment to the yaml that might get uncommented for syncing.

File: mlt/commands/init.py [hunk 14]
```diff
@@ -64,6 +65,24 @@ def action(self):
...
+
+                if self.args["--enable-sync"]:
+                    if localhost_helpers.binary_path('ksync'):
+                        # Syncthing uses '.stignore' to ignore files during
+                        # sync we also don't want to upload unneeded local data
+                        app_ignore_file = os.path.join(self.app_name,
+                                                       ".gitignore")
+                        ksync_ignore_file = os.path.join(self.app_name,
+                                                         ".stignore")
+                        copyfile(app_ignore_file, ksync_ignore_file)
```
[the remark this answers is not in the document]
Comment (ashahba) [reply mlt/commands/init.py]: Oh, so this one doesn't barf (just tried it and deleted `.gitignore` from template manually) because we have:
```
def copyfile(src, dst):
    """Copy data from src to dst"""
    if _samefile(src, dst):
        raise Error("`%s` and `%s` are the same file" % (src, dst))

    for fn in [src, dst]:
        try:
            st = os.stat(fn)
        except OSError:
            # File most likely does not exist
            pass
        else:
            # XXX What about other special files? (sockets, devices...)
            if stat.S_ISFIFO(st.st_mode):
                raise SpecialFileError("`%s` is a named pipe" % fn)

    with open(src, 'rb') as fsrc:
        with open(dst, 'wb') as fdst:
            copyfileobj(fsrc, fdst)
```
which doesn't error out if the file doesn't exist 🙂 

```
$ git status
On branch ashahba/mlt-sync-v2
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   README.md
	deleted:    mlt-templates/hello-world/.gitignore
	modified:   mlt/commands/init.py
	modified:   mlt/commands/sync.py
	modified:   mlt/main.py

no changes added to commit (use "git add" and/or "git commit -a")

$ mlt init hello-world-sync --template=hello-world --template-repo=. --enable-sync
[master (root-commit) 8c7bd2b] Initial commit.
 9 files changed, 149 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 .stignore
 create mode 100644 Dockerfile
 create mode 100644 Makefile
 create mode 100644 README.md
 create mode 100644 k8s-templates/job.yaml
 create mode 100644 k8s/.gitignore
 create mode 100644 main.py
 create mode 100644 requirements.txt


            Once your application is built and deployed try the following
            mlt commands:
            'mlt sync create' - to setup syncing local changes to the running
                           pods which in turn restarts the containers every
                           time changes are made to local template code.
                           This command only needs to run once.
            'mlt sync reload' - to wake up the 'sync' agent after a local
                           system reboot or long inactivity (default is 1 hour)
                           or any other activity that causes 'sync' agent to
                           die.
            'mlt sync delete' - to teardown syncing setup and stop syncing
                           local code changes with remote pods.
                           This command only need to run once.

            To ignore files and folders from syncing, add them to
            '.stignore' file.
            
```

File: mlt/commands/init.py [hunk 15]
```diff
@@ -79,6 +98,26 @@ def action(self):
...
 
+    def _update_yaml_for_sync(self):
+        # update k8s-template job spec, to keep the containers
+        # running so that we can update code locally and have ksync
+        # upload it to the running containers
+        k8s_template_dir = os.path.join(self.app_name,
+                                        "k8s-templates")
+        k8s_template_specs = []
+        for filename in os.listdir(k8s_template_dir):
+            if fnmatch.fnmatch(filename, '*job.yaml'):
```
Comment (s1113950) [inline mlt/commands/init.py]: This is gonna break `experiments.yaml` that was just added

File: mlt-templates/pytorch-distributed/k8s-templates/pytorchjob.yaml [hunk 16]
```diff
@@ -25,4 +28,7 @@ spec:
           - image: $image
             imagePullPolicy: IfNotPresent
             name: pytorch
+#           ### Uncomment this section to enable ksync
```
Comment (s1113950) [inline mlt-templates/pytorch-distributed/k8s-templates/pytorchjob.yaml]: I just realized this, this is basically the `patch_template_spec` function:
```
self.template_location['metadata'] = {'labels': {'debug': 'true'}}
        self.containers_location[0].update(
            {'command':
             ["/bin/bash", "-c", "trap : TERM INT; sleep infinity & wait"]})
        return json.dumps(data)
```
Can you just leverage this function so we don't have to have this in the yaml everywhere?

Comment (s1113950) [reply mlt-templates/pytorch-distributed/k8s-templates/pytorchjob.yaml]: you could rip that function, tweak it a bit, and put it in a util somewhere :)

File: docs/template_developers.md [hunk 17]
```diff
@@ -38,3 +38,20 @@ in to the `k8s-templates` directory when the app is deployed.
 parameter with `mlt templates list` and `mlt init` in order
 to specify the git url for  template repo (otherwise, this defaults to
 `https://github.com/IntelAI/mlt`).
+
+6. For templates supporting `mlt sync`, a couple of assumptions have been made:
+- The `Job` spec file under `k8s-templates`'s name must end with `job.yaml`.
```
Comment (dmsuehir) [inline docs/template_developers.md]: Just realized this was added when it was mentioned in the meeting.  Can't we just look for any `*.yaml` in k8s folder?  Why does it need to be named this.

File: tests/unit/commands/test_sync.py (hunk 6 above)
Comment (ashahba) [reply tests/unit/commands/test_sync.py]: Not sure who I can parameterize this one: `sync_output = sync(reload=True)`.
I have to think about this a little bit.

File: mlt/commands/init.py (hunk 15 above)
Comment (ashahba) [reply mlt/commands/init.py]: No since we don't have support for `ksync` on that template.

File: tests/unit/commands/test_sync.py (hunk 11 above)
Comment (ashahba) [reply tests/unit/commands/test_sync.py]: Not sure how I can parameterize this one: `sync_output = sync(reload=True)`.
I have to think about this a little bit.

File: mlt/commands/deploy.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ashahba) [reply mlt/commands/deploy.py]: @dmsuehir I resolved this now, we get a yellow colored warning message that says please run a command to manually delete the sync spec 🙂

**Comment by ashahba** [comment]
@s1113950 @dmsuehir if there are no more feedback left, let get this merged today and be done with it.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (dmsuehir) [reply README.md]: @ashahba Were you going to make this update to the README since you are splitting out required versus optional prereqs?
